### PR TITLE
Feat: Dry_run of a workflows

### PR DESCRIPTION
**Pull request by SebastianMorawiec**
# The problem
It would be useful to test the “decorators” of the tasks/workflows.

This could be done by running the workflow and each task is a no-op, but the resources, imports, etc. are all allocated/installed.
(https://zapatacomputing.atlassian.net/browse/ORQSDK-919)

# This PR's solution
Add new option to workflow runs that allows to run workflow without actually running the task code

# Checklist

_Check that this PR satisfies the following items:_

- [x] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [x] The PR's title is prefixed with `<feat/fix/chore/imp[rovement]/int[ernal]/docs>[!]:`
- [x] The PR is linked to a JIRA ticket (if there's no suitable ticket, check the box).

### CODE REVIEW

File: src/orquestra/sdk/_base/_driver/_models.py [hunk 1]
```diff
@@ -271,6 +271,7 @@ class CreateWorkflowRunRequest(pydantic.BaseModel):
 
     workflowDefinitionID: WorkflowDefID
     resources: Resources
+    dry_run: bool
```
Comment (alexjuda) [inline src/orquestra/sdk/_base/_driver/_models.py:274]: **issue**: Is it supported on the Platform already? This field isn't in the [CE's API spec](https://github.com/zapatacomputing/workflow-driver/blob/5d030e5a71be088c86590ce63180e256b55b3c8e/openapi/src/schemas/CreateWorkflowRunRequest.yaml). Even if we want to add client code before the Platform is ready, the spec should be updated first.

**Review by alexjuda — CHANGES_REQUESTED** [review]

File: src/orquestra/sdk/_base/_driver/_models.py (hunk 1 above)
Comment (SebastianMorawiec) [reply src/orquestra/sdk/_base/_driver/_models.py:274]: Its not supported - my intention is to wait for workflow-driver to support this feature before I merge this.

File: src/orquestra/sdk/_ray/_build_workflow.py [hunk 2]
```diff
@@ -152,7 +162,12 @@ def __call__(self, *wrapped_args, **wrapped_kwargs):
         for name, kwarg in wrapped_kwargs.items():
             kwargs[name] = self._unpack_argument(kwarg, name)
 
-        return self._user_fn(*args, **kwargs)
+        # if the dry_run mode is enabled, we don't execute user function, but just
+        # empty nop function returning Nones
+        if self._dry_run:
+            return self._nop_function_factory()()
+        else:
+            return self._user_fn(*args, **kwargs)
```
Comment (alexjuda) [inline src/orquestra/sdk/_ray/_build_workflow.py:170]: **nitpick**: this class (`ArgumentUnwrapper`) is about unwrapping arguments to make it suitable for interfacing with Ray tasks. The `dry_run` is an unrelated feature. Can you extract it to a different component? Functions are composable.

File: tests/runtime/ray/test_integration.py [hunk 3]
```diff
@@ -112,11 +112,34 @@ def test_sets_run_id_from_env(
...
+            run_id = runtime.create_workflow_run(wf_def, None, False)
 
             runtime.stop_workflow_run(run_id)
             assert run_id == global_wf_run_id
 
+        def test_simple_workflow_dry_run(
+            self, monkeypatch: pytest.MonkeyPatch, runtime: _dag.RayRuntime
+        ):
+            wf_def = _example_wfs.exception_wf_with_multiple_values.model
+            run_id = runtime.create_workflow_run(wf_def, None, True)
```
Comment (alexjuda) [inline tests/runtime/ray/test_integration.py:124]: **nitpick**: please use named arguments when passing literals.

File: tests/sdk/api/test_wf_run.py [hunk 4]
```diff
@@ -1366,7 +1370,7 @@ def test_run(
         monkeypatch.setattr(_api._config.RuntimeConfig, "name", "auto")
         with raises:
             wf_def.run("in_process", workspace_id=workspace_id, project_id=project_id)
-            workflow_create_mock.assert_called_once_with(wf_def.model, expected)
+            workflow_create_mock.assert_called_once_with(wf_def.model, expected, False)
```
Comment (alexjuda) [inline tests/sdk/api/test_wf_run.py:1373]: nitpick: please use named arguments when passing literals.

File: tests/sdk/driver/test_ce_runtime.py [hunk 5]
```diff
@@ -82,7 +82,7 @@ def test_happy_path(
         mocked_client.create_workflow_run.return_value = workflow_run_id
 
         # When
-        wf_run_id = runtime.create_workflow_run(my_workflow.model, None)
+        wf_run_id = runtime.create_workflow_run(my_workflow.model, None, False)
```
Comment (alexjuda) [inline tests/sdk/driver/test_ce_runtime.py:85]: nitpick: please use named arguments when passing literals.

File: tests/sdk/driver/test_client.py [hunk 6]
```diff
@@ -1040,7 +1040,7 @@ def test_invalid_request(
                 )
 
                 with pytest.raises(_exceptions.InvalidWorkflowRunRequest):
-                    _ = client.create_workflow_run(workflow_def_id, resources)
+                    _ = client.create_workflow_run(workflow_def_id, resources, False)
```
Comment (alexjuda) [inline tests/sdk/driver/test_client.py:1043]: nitpick: please use named arguments when passing literals.

File: tests/sdk/test_in_process_runtime.py [hunk 7]
```diff
@@ -77,14 +77,16 @@ def test_secret_inside_ir(
     get_secret.return_value = mocked_secret
     monkeypatch.setattr(_client.SecretsClient, "get_secret", get_secret)
 
-    run_id = runtime.create_workflow_run(wf_with_secrets().model, None)
+    run_id = runtime.create_workflow_run(wf_with_secrets().model, None, False)
```
Comment (alexjuda) [inline tests/sdk/test_in_process_runtime.py:80]: nitpick: please use named arguments when passing literals.

File: tests/runtime/ray/test_build_workflow.py [hunk 8]
```diff
@@ -391,7 +399,7 @@ def test_positional(
         ):
             fn = Mock()
             arg_unwrapper = _build_workflow.ArgumentUnwrapper(
-                fn, args_artifact_nodes, {}, True
+                fn, args_artifact_nodes, {}, True, 1, False
```
Comment (alexjuda) [inline tests/runtime/ray/test_build_workflow.py:402]: nitpick: please use named arguments when passing literals.

**Review by alexjuda — CHANGES_REQUESTED** [review]

File: tests/sdk/api/test_wf_run.py (hunk 4 above)
Comment (SebastianMorawiec) [reply tests/sdk/api/test_wf_run.py:1373]: For all the comments about passing literals: I can do changes with function call, cant apply them to mock.asserts - if parameter is passed as arg, it has to be checked as arg, not kwarg

File: src/orquestra/sdk/_ray/_build_workflow.py [hunk 9]
```diff
@@ -142,6 +146,12 @@ def _unpack_argument(
         else:
             assert_never(arg)
 
+    def _nop_function_factory(self):
+        def nop():
+            return (None,) * self._n_outputs
```
[the remark this answers is not in the document]
Comment (SebastianMorawiec) [reply src/orquestra/sdk/_ray/_build_workflow.py:151]: Ive also checked your example of:
```
@sdk.task
def foo():
    return "abc"
```

- Right now we treat strings as non-subscribtable. So return "Abc" is the same for us as return <any literal like None>

File: src/orquestra/sdk/_ray/_build_workflow.py [hunk 10]
```diff
@@ -426,6 +453,8 @@ def make_ray_dag(
             n_outputs=_compat.n_outputs(task_def=user_task, task_inv=invocation),
             project_dir=project_dir,
             user_fn_ref=user_task.fn_ref,
+            output_metadata=user_task.output_metadata,
```
Comment (alexjuda) [inline src/orquestra/sdk/_ray/_build_workflow.py:456]: **nitpick**: shouldn't `n_outputs` be replaced with this parameter?

**Review by alexjuda — APPROVED** [review]

**Pull request merged**